RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+    // it is important to remember that the actual rank might be smaller 
+    // than the currently-defined one, so this should not have an early termination.
+    if((id < indicator) && (std::abs(A[id * lda + id]) / std::abs(A[0]) <= alpha)) {
+        atomicMin((int*) &indicator, (int) id);
+    }
+    __syncthreads();
+    if(id == 0 && indicator != n) {
+        *rank = indicator;
+    }
+}
+
+template <typename T>
+__global__ void __launch_bounds__(128) elementwise_product(
+int64_t n,
+const T alpha,
+const T* A,
+int64_t lda,
+const T* B,
+int64_t ldb,
+T* C,
+int64_t ldc
+){
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        C[id * ldc] = alpha * A[id * lda] * B[id * ldb];
+    }
+}
+
+
+template <typename T>
+__global__ void __launch_bounds__(1) diagonal_update(
+    int i,
+    T* A,
+    int64_t lda,
+    T* D,
+    int64_t n
+){
+    if (threadIdx.x != 0) {
+        return;
+    }
+    // S(i, i) = − sgn(Q(i, i)); = 1 if Q(i, i) == 0
+    T buf = A[i * lda + i];
+    T new_d = (buf == 0) ? 1 : -((buf > T{0}) - (buf < T{0}));
+    A[i * lda + i] -= new_d;
+    D[i] = new_d;
+}
+
+template <typename T>
+__global__ void __launch_bounds__(128) copy_gpu(
+    int64_t n, 
+    T const* src, 
+    int64_t incsrc, 
+    T* dest, 
+    int64_t incdest
+) {
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        dest[id * incdest] = src[id * incsrc];
+    }
+}
+
+template <typename T>
+__global__ void __launch_bounds__(128) fill_gpu(
+    int64_t n, 
+    T* x, 
+    int64_t incx, 
+    T alpha
+) {
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        x[id * incx] = alpha;
+    }
+}
+
+template <typename T>
+__global__ void __launch_bounds__(128) fill_gpu(
+    int64_t n, 
+    int64_t* x, 
+    int64_t incx, 
+    T alpha
+) {
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        x[id * incx] = alpha;
+    }
+}
+
+template <typename T>
+__device__ void copy_gpu_device(
+    int64_t n,
+    T const* src,
+    int64_t incsrc,
+    T* dest,
+    int64_t incdest
+) {
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        dest[id * incdest] = src[id * incsrc];
+    }
+}
+
+template <typename T>
+__global__ void __launch_bounds__(128) swap_gpu(
+    T* a, 
+    int64_t inca, 
+    T* b, 
+    int64_t n, 
+    int64_t incb
+) {
+    int64_t const id = (int64_t)blockDim.x * blockIdx.x + threadIdx.x;
+    if (id < n) {
+        T const v{a[id * inca]};
+        a[id * inca] = b[id * incb];
+        b[id * incb] = v;
+    }
+}
+
+// This operation is actually not parallelizable, as the vector b may contain repeated indices 
+// that need to be accessed in a sepcific order. 
+template <typename Idx>
+__global__ void __launch_bounds__(128) LUQRCP_piv_porcess_gpu_global(
```

Comment:
Please add a comment explaining what this function does. I think I know, but I'm not sure.

Also, in the function name, "porcess" --> "process."

Reply:
Transforms the LU pivot format into QR pivot format, used in the context of LU-based QRCP.
Fixed.